// graphile-build/dist/plugins/AddNodeInterfaceToSuitableTypesPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddNodeInterfaceToSuitableTypesPlugin = void 0;
require("graphile-config");
const grafast_1 = require("grafast");
const utils_js_1 = require("../utils.js");
const NodePlugin_js_1 = require("./NodePlugin.js");
exports.AddNodeInterfaceToSuitableTypesPlugin = {
    name: "AddNodeInterfaceToSuitableTypesPlugin",
    version: "1.0.0",
    description: `Adds the 'Node' interface to all types that have registered a Node ID handler`,
    schema: {
        hooks: {
            GraphQLObjectType_interfaces(interfaces, build, context) {
                const { getTypeByName, inflection, [NodePlugin_js_1.NODE_ID_HANDLER_BY_TYPE_NAME]: nodeIdHandlerByTypeName, } = build;
                if (!nodeIdHandlerByTypeName) {
                    return interfaces;
                }
                const { Self } = context;
                const handler = nodeIdHandlerByTypeName[Self.name];
                if (handler == null) {
                    return interfaces;
                }
                const Type = getTypeByName(inflection.builtin("Node"));
                if (!Type) {
                    return interfaces;
                }
                return [...interfaces, Type];
            },
            GraphQLObjectType_fields(fields, build, context) {
                const { getTypeByName, inflection, [NodePlugin_js_1.NODE_ID_HANDLER_BY_TYPE_NAME]: nodeIdHandlerByTypeName, [NodePlugin_js_1.NODE_ID_CODECS]: nodeIdCodecs, graphql: { GraphQLNonNull, GraphQLID }, } = build;
                if (!nodeIdHandlerByTypeName) {
                    return fields;
                }
                const { Self, scope: { isRootQuery }, } = context;
                const handler = nodeIdHandlerByTypeName[Self.name];
                if (handler == null) {
                    return fields;
                }
                const Type = getTypeByName(inflection.builtin("Node"));
                if (!Type) {
                    return fields;
                }
                return build.extend(fields, {
                    [build.inflection.nodeIdFieldName()]: {
                        description: isRootQuery
                            ? build.wrapDescription("The root query type must be a `Node` to work well with Relay 1 mutations. This just resolves to `query`.", "field")
                            : build.wrapDescription("A globally unique identifier. Can be used in various places throughout the system to identify this single value.", "field"),
                        type: new GraphQLNonNull(GraphQLID),
                        plan: (0, utils_js_1.EXPORTABLE)((handler, lambda, nodeIdCodecs) => ($parent) => {
                            const specifier = handler.plan($parent);
                            return lambda(specifier, nodeIdCodecs[handler.codecName].encode);
                        }, [handler, grafast_1.lambda, nodeIdCodecs]),
                    },
                }, "Adding id field to Query type from AddNodeInterfaceToQueryPlugin");
            },
        },
    },
};
