// graphile-build/dist/plugins/CursorTypePlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CursorTypePlugin = void 0;
require("graphile-config");
const utils_js_1 = require("../utils.js");
const version_js_1 = require("../version.js");
exports.CursorTypePlugin = {
    name: "CursorTypePlugin",
    description: "Registers the 'Cursor' scalar type for cursor pagination",
    version: version_js_1.version,
    schema: {
        hooks: {
            init: {
                callback: (_, build) => {
                    const { registerScalarType, inflection } = build;
                    registerScalarType(inflection.builtin("Cursor"), { isCursorType: true }, () => ({
                        ...utils_js_1.stringScalarSpec,
                        description: "A location in a connection that can be used for resuming pagination.",
                    }), "graphile-build built-in (Cursor type)");
                    return _;
                },
                provides: ["Cursor"],
            },
        },
    },
};
