// graphile-build/dist/plugins/RegisterQueryNodePlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegisterQueryNodePlugin = void 0;
require("graphile-config");
const grafast_1 = require("grafast");
const utils_js_1 = require("../utils.js");
exports.RegisterQueryNodePlugin = {
    name: "RegisterQueryNodePlugin",
    version: "1.0.0",
    description: `Registers the 'Query' type as a 'Node' type. You probably don't want this.`,
    schema: {
        hooks: {
            init(_, build) {
                if (!build.registerNodeIdHandler) {
                    return _;
                }
                build.registerNodeIdHandler({
                    typeName: build.inflection.builtin("Query"),
                    codecName: "raw",
                    match: (0, utils_js_1.EXPORTABLE)(() => (specifier) => {
                        return specifier === "query";
                    }, []),
                    getSpec: () => "irrelevant",
                    get: (0, utils_js_1.EXPORTABLE)((operationPlan) => () => {
                        return operationPlan().rootValueStep;
                    }, [grafast_1.operationPlan]),
                    plan: (0, utils_js_1.EXPORTABLE)((constant) => () => {
                        return constant `query`;
                    }, [grafast_1.constant]),
                });
                return _;
            },
        },
    },
};
