// graphile-build-pg/dist/plugins/PgMutationUpdateDeletePlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PgMutationUpdateDeletePlugin = void 0;
const tslib_1 = require("tslib");
require("graphile-config");
const pg_1 = require("@dataplan/pg");
const grafast_1 = require("grafast");
const graphile_build_1 = require("graphile-build");
const tamedevil_1 = tslib_1.__importStar(require("tamedevil"));
const utils_js_1 = require("../utils.js");
const version_js_1 = require("../version.js");
const isUpdatable = (build, resource) => {
    if (resource.parameters)
        return false;
    if (!resource.codec.attributes)
        return false;
    if (resource.codec.polymorphism)
        return false;
    if (resource.codec.isAnonymous)
        return false;
    if (!resource.uniques || resource.uniques.length < 1)
        return false;
    return !!build.behavior.pgResourceMatches(resource, "resource:update");
};
const isDeletable = (build, resource) => {
    if (resource.parameters)
        return false;
    if (!resource.codec.attributes)
        return false;
    if (resource.codec.polymorphism)
        return false;
    if (resource.codec.isAnonymous)
        return false;
    if (!resource.uniques || resource.uniques.length < 1)
        return false;
    return !!build.behavior.pgResourceMatches(resource, "resource:delete");
};
exports.PgMutationUpdateDeletePlugin = {
    name: "PgMutationUpdateDeletePlugin",
    description: "Adds 'update' and 'delete' mutations for supported sources",
    version: version_js_1.version,
    after: ["smart-tags", "PgTablesPlugin", "PgCodecsPlugin", "PgTypesPlugin"],
    inflection: {
        add: {
            updatePayloadType(options, { resource }) {
                return this.upperCamelCase(`update-${this._singularizedResourceName(resource)}-payload`);
            },
            deletePayloadType(options, { resource }) {
                return this.upperCamelCase(`delete-${this._singularizedResourceName(resource)}-payload`);
            },
            updateNodeField(options, { resource, unique: _unique }) {
                return this.camelCase(`update-${this._singularizedResourceName(resource)}`);
            },
            updateNodeInputType(options, details) {
                return this.upperCamelCase(`${this.updateNodeField(details)}-input`);
            },
            deletedNodeId(options, { resource }) {
                return this.camelCase(`deleted-${this._singularizedResourceName(resource)}-${this.nodeIdFieldName()}`);
            },
            deleteNodeField(options, { resource, unique: _unique }) {
                return this.camelCase(`delete-${this._singularizedResourceName(resource)}`);
            },
            deleteNodeInputType(options, details) {
                return this.upperCamelCase(`${this.deleteNodeField(details)}-input`);
            },
            updateByKeysField(options, { resource, unique }) {
                return this.camelCase(`update-${this._singularizedResourceName(resource)}-by-${this._joinAttributeNames(resource.codec, unique.attributes)}`);
            },
            updateByKeysInputType(options, details) {
                return this.upperCamelCase(`${this.updateByKeysField(details)}-input`);
            },
            deleteByKeysField(options, { resource, unique }) {
                return this.camelCase(`delete-${this._singularizedResourceName(resource)}-by-${this._joinAttributeNames(resource.codec, unique.attributes)}`);
            },
            deleteByKeysInputType(options, details) {
                return this.upperCamelCase(`${this.deleteByKeysField(details)}-input`);
            },
            patchField(options, fieldName) {
                return this.camelCase(`${fieldName}-patch`);
            },
        },
    },
    schema: {
        entityBehavior: {
            pgResource: "update delete",
            pgResourceUnique: "update delete",
        },
        hooks: {
            init(_, build) {
                const { inflection, graphql: { GraphQLString, GraphQLNonNull, GraphQLID }, } = build;
                const process = (resource, mode) => {
                    const modeText = mode === "resource:update" ? "update" : "delete";
                    const tableTypeName = inflection.tableType(resource.codec);
                    const payloadTypeName = mode === "resource:update"
                        ? inflection.updatePayloadType({ resource })
                        : inflection.deletePayloadType({ resource });
                    // Payload type is shared independent of the keys used
                    build.registerObjectType(payloadTypeName, {
                        isMutationPayload: true,
                        isPgUpdatePayloadType: mode === "resource:update",
                        isPgDeletePayloadType: mode === "resource:delete",
                        pgTypeResource: resource,
                    }, grafast_1.ObjectStep, () => {
                        return {
                            description: build.wrapDescription(`The output of our ${modeText} \`${tableTypeName}\` mutation.`, "type"),
                            fields: ({ fieldWithHooks }) => {
                                const tableName = inflection.tableFieldName(resource);
                                const deletedNodeIdFieldName = build.getNodeIdHandler !== undefined
                                    ? inflection.deletedNodeId({
                                        resource,
                                    })
                                    : null;
                                const TableType = build.getGraphQLTypeByPgCodec(resource.codec, "output");
                                const handler = TableType && build.getNodeIdHandler
                                    ? build.getNodeIdHandler(TableType.name)
                                    : null;
                                const nodeIdCodec = handler
                                    ? build.getNodeIdCodec(handler.codecName)
                                    : null;
                                return {
                                    clientMutationId: {
                                        description: build.wrapDescription("The exact same `clientMutationId` that was provided in the mutation input, unchanged and unused. May be used by a client to track mutations.", "field"),
                                        type: GraphQLString,
                                        plan: (0, graphile_build_1.EXPORTABLE)((constant) => function plan($mutation) {
                                            return ($mutation.getStepForKey("clientMutationId", true) ?? constant(null));
                                        }, [grafast_1.constant]),
                                    },
                                    // TODO: default to `...(mode === "resource:update" && TableType`; we only want the record on delete for v4 compatibility
                                    ...(TableType
                                        ? {
                                            [tableName]: fieldWithHooks({
                                                fieldName: tableName,
                                                fieldBehaviorScope: `update:payload:record`,
                                            }, () => ({
                                                description: build.wrapDescription(`The \`${tableTypeName}\` that was ${modeText}d by this mutation.`, "field"),
                                                type: TableType,
                                                plan: (0, graphile_build_1.EXPORTABLE)(() => function plan($object) {
                                                    return $object.get("result");
                                                }, []),
                                            })),
                                        }
                                        : {}),
                                    ...(mode === "resource:delete" &&
                                        deletedNodeIdFieldName &&
                                        handler &&
                                        nodeIdCodec &&
                                        build.behavior.pgResourceMatches(resource, "node")
                                        ? {
                                            [deletedNodeIdFieldName]: fieldWithHooks({
                                                fieldName: deletedNodeIdFieldName,
                                                // TODO: fieldBehaviorScope: `...`,
                                                isPgMutationPayloadDeletedNodeIdField: true,
                                            }, () => {
                                                return {
                                                    type: GraphQLID,
                                                    plan: (0, graphile_build_1.EXPORTABLE)((handler, lambda, nodeIdCodec) => function plan($object) {
                                                        const $record = $object.getStepForKey("result");
                                                        const specifier = handler.plan($record);
                                                        return lambda(specifier, nodeIdCodec.encode);
                                                    }, [handler, grafast_1.lambda, nodeIdCodec]),
                                                };
                                            }),
                                        }
                                        : null),
                                };
                            },
                        };
                    }, `Creating ${mode} payload for ${resource} from PgMutationUpdateDeletePlugin`);
                    const primaryUnique = resource.uniques.find((u) => u.isPrimary);
                    const specs = [
                        ...(primaryUnique && build.getNodeIdCodec !== undefined
                            ? [{ unique: primaryUnique, uniqueMode: "node" }]
                            : []),
                        ...resource.uniques.map((unique) => ({
                            unique,
                            uniqueMode: "keys",
                        })),
                    ];
                    for (const spec of specs) {
                        const { uniqueMode, unique } = spec;
                        const details = {
                            resource,
                            unique,
                        };
                        if (uniqueMode === "node" && !build.getNodeIdHandler) {
                            continue;
                        }
                        build.recoverable(null, () => {
                            const tablePatchName = build.getGraphQLTypeNameByPgCodec(resource.codec, "patch");
                            if (!tablePatchName && mode === "resource:update") {
                                return;
                            }
                            const inputTypeName = mode === "resource:update"
                                ? uniqueMode === "node"
                                    ? inflection.updateNodeInputType(details)
                                    : inflection.updateByKeysInputType(details)
                                : uniqueMode === "node"
                                    ? inflection.deleteNodeInputType(details)
                                    : inflection.deleteByKeysInputType(details);
                            const fieldName = mode === "resource:update"
                                ? uniqueMode === "node"
                                    ? inflection.updateNodeField(details)
                                    : inflection.updateByKeysField(details)
                                : uniqueMode === "node"
                                    ? inflection.deleteNodeField(details)
                                    : inflection.deleteByKeysField(details);
                            const nodeIdFieldName = uniqueMode === "node" ? inflection.nodeIdFieldName() : null;
                            build.registerInputObjectType(inputTypeName, {
                                isPgUpdateInputType: mode === "resource:update",
                                isPgUpdateByKeysInputType: mode === "resource:update" && uniqueMode === "keys",
                                isPgUpdateNodeInputType: mode === "resource:update" && uniqueMode === "node",
                                isPgDeleteInputType: mode === "resource:delete",
                                isPgDeleteByKeysInputType: mode === "resource:delete" && uniqueMode === "keys",
                                isPgDeleteNodeInputType: mode === "resource:delete" && uniqueMode === "node",
                                pgResource: resource,
                                pgResourceUnique: unique,
                                isMutationInput: true,
                            }, () => {
                                const TablePatch = mode === "resource:update"
                                    ? build.getInputTypeByName(tablePatchName)
                                    : null;
                                return {
                                    description: build.wrapDescription(`All input for the \`${fieldName}\` mutation.`, "type"),
                                    fields: Object.assign({
                                        clientMutationId: {
                                            description: build.wrapDescription("An arbitrary string value with no semantic meaning. Will be included in the payload verbatim. May be used to track mutations by the client.", "field"),
                                            type: GraphQLString,
                                            applyPlan: (0, graphile_build_1.EXPORTABLE)(() => function plan($input, val) {
                                                $input.set("clientMutationId", val.get());
                                            }, []),
                                        },
                                        ...(uniqueMode === "node"
                                            ? {
                                                [nodeIdFieldName]: {
                                                    description: build.wrapDescription(`The globally unique \`ID\` which will identify a single \`${tableTypeName}\` to be ${modeText}d.`, "field"),
                                                    type: new GraphQLNonNull(GraphQLID),
                                                },
                                            }
                                            : unique.attributes.reduce((memo, attributeName) => {
                                                const attribute = resource.codec.attributes[attributeName];
                                                memo[inflection.attribute({
                                                    attributeName,
                                                    codec: resource.codec,
                                                })] = {
                                                    description: attribute.description,
                                                    type: new GraphQLNonNull(build.getGraphQLTypeByPgCodec(attribute.codec, "input")),
                                                };
                                                return memo;
                                            }, Object.create(null))),
                                    }, mode === "resource:update"
                                        ? {
                                            [inflection.patchField(inflection.tableFieldName(resource))]: {
                                                description: build.wrapDescription(`An object where the defined keys will be set on the \`${tableTypeName}\` being ${modeText}d.`, "field"),
                                                type: new GraphQLNonNull(TablePatch),
                                                applyPlan: (0, graphile_build_1.EXPORTABLE)(() => function plan($object) {
                                                    const $record = $object.getStepForKey("result");
                                                    return $record.setPlan();
                                                }, []),
                                            },
                                        }
                                        : null),
                                };
                            }, `Creating ${mode} input by ${uniqueMode} for ${unique.attributes.join(",")} of ${resource} from PgMutationUpdateDeletePlugin`);
                        });
                    }
                };
                const allResources = Object.values(build.input.pgRegistry.pgResources);
                const updatableResources = allResources.filter((resource) => isUpdatable(build, resource));
                const deletableResources = allResources.filter((resource) => isDeletable(build, resource));
                updatableResources.forEach((resource) => {
                    process(resource, "resource:update");
                });
                deletableResources.forEach((resource) => {
                    process(resource, "resource:delete");
                });
                return _;
            },
            GraphQLObjectType_fields(fields, build, context) {
                const { inflection, graphql: { GraphQLNonNull }, } = build;
                const { scope: { isRootMutation }, fieldWithHooks, } = context;
                const nodeIdFieldName = build.inflection.nodeIdFieldName?.();
                if (!isRootMutation) {
                    return fields;
                }
                const allSources = Object.values(build.input.pgRegistry.pgResources);
                const updatableSources = allSources.filter((resource) => isUpdatable(build, resource));
                const deletableSources = allSources.filter((resource) => isDeletable(build, resource));
                const process = (fields, resources, mode) => {
                    for (const resource of resources) {
                        const payloadTypeName = mode === "resource:update"
                            ? inflection.updatePayloadType({ resource })
                            : inflection.deletePayloadType({ resource });
                        const primaryUnique = resource.uniques.find((u) => u.isPrimary);
                        const constraintMode = `constraint:${mode}`;
                        const specs = [
                            ...(primaryUnique && !!build.getNodeIdHandler
                                ? [{ unique: primaryUnique, uniqueMode: "node" }]
                                : []),
                            ...resource.uniques.map((unique) => ({
                                unique,
                                uniqueMode: "keys",
                            })),
                        ].filter((spec) => {
                            const unique = spec.unique;
                            return !!build.behavior.pgResourceUniqueMatches([resource, unique], constraintMode);
                        });
                        for (const spec of specs) {
                            const { uniqueMode, unique } = spec;
                            const details = {
                                resource,
                                unique,
                            };
                            fields = build.recoverable(fields, () => {
                                const fieldName = mode === "resource:update"
                                    ? uniqueMode === "node"
                                        ? inflection.updateNodeField(details)
                                        : inflection.updateByKeysField(details)
                                    : uniqueMode === "node"
                                        ? inflection.deleteNodeField(details)
                                        : inflection.deleteByKeysField(details);
                                const inputTypeName = mode === "resource:update"
                                    ? uniqueMode === "node"
                                        ? inflection.updateNodeInputType(details)
                                        : inflection.updateByKeysInputType(details)
                                    : uniqueMode === "node"
                                        ? inflection.deleteNodeInputType(details)
                                        : inflection.deleteByKeysInputType(details);
                                const payloadType = build.getOutputTypeByName(payloadTypeName);
                                const mutationInputType = build.getTypeByName(inputTypeName);
                                if (!mutationInputType) {
                                    return fields;
                                }
                                if (!build.graphql.isInputObjectType(mutationInputType)) {
                                    throw new Error(`Expected '${inputTypeName}' to be an input object type`);
                                }
                                const uniqueAttributes = unique.attributes.map((attributeName) => [
                                    attributeName,
                                    inflection.attribute({
                                        attributeName,
                                        codec: resource.codec,
                                    }),
                                ]);
                                /**
                                 * If every attribute is a safe identifier then we can create an
                                 * optimised function, otherwise we must play it safe and not
                                 * do that.
                                 */
                                const clean = uniqueMode === "keys" &&
                                    uniqueAttributes.every(([attributeName, fieldName]) => (0, tamedevil_1.isSafeObjectPropertyName)(attributeName) &&
                                        (0, tamedevil_1.isSafeObjectPropertyName)(fieldName));
                                /**
                                 * Builds a pgUpdateSingle/pgDeleteSingle spec describing the row to
                                 * update/delete as a string containing raw JS code if it's
                                 * safe to do so. This enables us to create an optimised
                                 * function for the plan resolver, especially good for the
                                 * exported schema.
                                 */
                                const specFromArgsString = clean
                                    ? (0, tamedevil_1.default) `{ ${tamedevil_1.default.join(uniqueAttributes.map(([attributeName, fieldName]) => (0, tamedevil_1.default) `${tamedevil_1.default.dangerousKey(attributeName)}: args.get(['input', ${tamedevil_1.default.lit(fieldName)}])`), ", ")} }`
                                    : null;
                                const tableTypeName = inflection.tableType(resource.codec);
                                const handler = build.getNodeIdHandler
                                    ? build.getNodeIdHandler(tableTypeName)
                                    : null;
                                const codec = handler
                                    ? build.getNodeIdCodec(handler.codecName)
                                    : null;
                                if (uniqueMode !== "keys" && (!codec || !handler)) {
                                    return fields;
                                }
                                /**
                                 * The fallback to `specFromArgsString`; builds a
                                 * pgUpdateSingle/pgDeleteSingle spec describing the row to update/delete.
                                 */
                                const specFromArgs = uniqueMode === "keys"
                                    ? (0, graphile_build_1.EXPORTABLE)((uniqueAttributes) => (args) => {
                                        return uniqueAttributes.reduce((memo, [attributeName, fieldName]) => {
                                            memo[attributeName] = args.get([
                                                "input",
                                                fieldName,
                                            ]);
                                            return memo;
                                        }, Object.create(null));
                                    }, [uniqueAttributes])
                                    : (0, graphile_build_1.EXPORTABLE)((codec, handler, nodeIdFieldName, specFromNodeId) => (args) => {
                                        const $nodeId = args.get([
                                            "input",
                                            nodeIdFieldName,
                                        ]);
                                        return specFromNodeId(codec, handler, $nodeId);
                                    }, [codec, handler, nodeIdFieldName, grafast_1.specFromNodeId]);
                                return build.extend(fields, {
                                    [fieldName]: fieldWithHooks({ fieldName, fieldBehaviorScope: constraintMode }, {
                                        args: {
                                            input: {
                                                type: new GraphQLNonNull(mutationInputType),
                                                applyPlan: (0, graphile_build_1.EXPORTABLE)(() => function plan(_, $object) {
                                                    return $object;
                                                }, []),
                                            },
                                        },
                                        type: payloadType,
                                        description: `${mode === "resource:update" ? "Updates" : "Deletes"} a single \`${inflection.tableType(resource.codec)}\` ${uniqueMode === "keys"
                                            ? "using a unique key"
                                            : "using its globally unique id"}${mode === "resource:update" ? " and a patch" : ""}.`,
                                        deprecationReason: (0, utils_js_1.tagToString)(resource.extensions?.tags?.deprecated),
                                        plan: mode === "resource:update"
                                            ? specFromArgsString
                                                ? // eslint-disable-next-line graphile-export/exhaustive-deps
                                                    (0, graphile_build_1.EXPORTABLE)(tamedevil_1.default.run `\
return function(object, pgUpdateSingle, resource) {
return (_$root, args) => {
  const plan = object({ result: pgUpdateSingle(resource, ${specFromArgsString}) });
  args.apply(plan);
  return plan;
}
}`, [grafast_1.object, pg_1.pgUpdateSingle, resource])
                                                : (0, graphile_build_1.EXPORTABLE)((object, pgUpdateSingle, resource, specFromArgs) => function plan(_$root, args) {
                                                    const plan = object({
                                                        result: pgUpdateSingle(resource, specFromArgs(args)),
                                                    });
                                                    args.apply(plan);
                                                    return plan;
                                                }, [
                                                    grafast_1.object,
                                                    pg_1.pgUpdateSingle,
                                                    resource,
                                                    specFromArgs,
                                                ])
                                            : specFromArgsString
                                                ? // eslint-disable-next-line graphile-export/exhaustive-deps
                                                    (0, graphile_build_1.EXPORTABLE)(tamedevil_1.default.run `\
return function (object, pgDeleteSingle, resource) {
return (_$root, args) => {
  const plan = object({ result: pgDeleteSingle(resource, ${specFromArgsString}) });
  args.apply(plan);
  return plan;
}
}`, [grafast_1.object, pg_1.pgDeleteSingle, resource])
                                                : (0, graphile_build_1.EXPORTABLE)((object, pgDeleteSingle, resource, specFromArgs) => function plan(_$root, args) {
                                                    const plan = object({
                                                        result: pgDeleteSingle(resource, specFromArgs(args)),
                                                    });
                                                    args.apply(plan);
                                                    return plan;
                                                }, [
                                                    grafast_1.object,
                                                    pg_1.pgDeleteSingle,
                                                    resource,
                                                    specFromArgs,
                                                ]),
                                    }),
                                }, `Adding ${mode} mutation for ${resource}`);
                            });
                        }
                    }
                    return fields;
                };
                return process(process(fields, updatableSources, "resource:update"), deletableSources, "resource:delete");
            },
        },
    },
};
