// graphile-build-pg/dist/plugins/PgOrderByPrimaryKeyPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PgOrderByPrimaryKeyPlugin = void 0;
require("./PgTablesPlugin.js");
require("graphile-config");
const graphile_build_1 = require("graphile-build");
const version_js_1 = require("../version.js");
exports.PgOrderByPrimaryKeyPlugin = {
    name: "PgOrderByPrimaryKeyPlugin",
    description: "Adds ordering by the table's primary key",
    version: version_js_1.version,
    schema: {
        hooks: {
            GraphQLEnumType_values(values, build, context) {
                const { extend, inflection, sql, options } = build;
                const { scope: { isPgRowSortEnum, pgCodec: rawPgCodec }, } = context;
                const { pgOrderByNullsLast } = options;
                if (!isPgRowSortEnum ||
                    !rawPgCodec ||
                    !rawPgCodec.attributes ||
                    rawPgCodec.isAnonymous) {
                    return values;
                }
                const pgCodec = rawPgCodec;
                const resources = Object.values(build.input.pgRegistry.pgResources).filter((s) => s.codec === pgCodec && !s.parameters);
                if (resources.length < 1) {
                    return values;
                }
                const primaryKey = resources[0].uniques.find((resource) => resource.isPrimary);
                if (!primaryKey) {
                    return values;
                }
                const primaryKeyAttributes = primaryKey.attributes;
                return extend(values, {
                    [inflection.builtin("PRIMARY_KEY_ASC")]: {
                        extensions: {
                            grafast: {
                                applyPlan: (0, graphile_build_1.EXPORTABLE)((pgCodec, pgOrderByNullsLast, primaryKeyAttributes, sql) => (step) => {
                                    primaryKeyAttributes.forEach((attributeName) => {
                                        const attribute = pgCodec.attributes[attributeName];
                                        step.orderBy({
                                            codec: attribute.codec,
                                            fragment: sql `${step.alias}.${sql.identifier(attributeName)}`,
                                            direction: "ASC",
                                            ...(pgOrderByNullsLast != null
                                                ? {
                                                    nulls: pgOrderByNullsLast ? "LAST" : "FIRST",
                                                }
                                                : null),
                                        });
                                    });
                                    step.setOrderIsUnique();
                                }, [pgCodec, pgOrderByNullsLast, primaryKeyAttributes, sql]),
                            },
                        },
                    },
                    [inflection.builtin("PRIMARY_KEY_DESC")]: {
                        extensions: {
                            grafast: {
                                applyPlan: (0, graphile_build_1.EXPORTABLE)((pgCodec, pgOrderByNullsLast, primaryKeyAttributes, sql) => (step) => {
                                    primaryKeyAttributes.forEach((attributeName) => {
                                        const attribute = pgCodec.attributes[attributeName];
                                        step.orderBy({
                                            codec: attribute.codec,
                                            fragment: sql `${step.alias}.${sql.identifier(attributeName)}`,
                                            direction: "DESC",
                                            ...(pgOrderByNullsLast != null
                                                ? {
                                                    nulls: pgOrderByNullsLast ? "LAST" : "FIRST",
                                                }
                                                : null),
                                        });
                                    });
                                    step.setOrderIsUnique();
                                }, [pgCodec, pgOrderByNullsLast, primaryKeyAttributes, sql]),
                            },
                        },
                    },
                }, `Adding primary key asc/desc sort to table '${pgCodec.name}'`);
            },
        },
    },
};
