// pg-introspection/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseIntrospectionResults = exports.resolvePermissions = exports.expandRoles = exports.aclContainsRole = exports.parseSmartComment = exports.makeIntrospectionQuery = void 0;
var introspection_js_1 = require("./introspection.js");
Object.defineProperty(exports, "makeIntrospectionQuery", { enumerable: true, get: function () { return introspection_js_1.makeIntrospectionQuery; } });
const acl_js_1 = require("./acl.js");
Object.defineProperty(exports, "aclContainsRole", { enumerable: true, get: function () { return acl_js_1.aclContainsRole; } });
Object.defineProperty(exports, "expandRoles", { enumerable: true, get: function () { return acl_js_1.expandRoles; } });
Object.defineProperty(exports, "resolvePermissions", { enumerable: true, get: function () { return acl_js_1.resolvePermissions; } });
const augmentIntrospection_js_1 = require("./augmentIntrospection.js");
var smartComments_js_1 = require("./smartComments.js");
Object.defineProperty(exports, "parseSmartComment", { enumerable: true, get: function () { return smartComments_js_1.parseSmartComment; } });
function parseIntrospectionResults(introspectionResults, includeExtensionResources = false) {
    return (0, augmentIntrospection_js_1.augmentIntrospection)(introspectionResults, includeExtensionResources);
}
exports.parseIntrospectionResults = parseIntrospectionResults;
