// grafserv/dist/middleware/graphiql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeGraphiQLHandler = void 0;
const server_1 = require("ruru/server");
const hooks_js_1 = require("../hooks.js");
// TODO: use a specific version of mermaid
function makeGraphiQLHandler(resolvedPreset, dynamicOptions) {
    const { htmlParts: htmlPartsFromConfig } = resolvedPreset?.ruru ?? {};
    const hooks = (0, hooks_js_1.getGrafservHooks)(resolvedPreset);
    const unhookedHTMLParts = {
        ...server_1.defaultHTMLParts,
        ...htmlPartsFromConfig,
    };
    return async (request) => {
        let htmlParts = unhookedHTMLParts;
        if (hooks.callbacks.ruruHTMLParts) {
            htmlParts = {
                ...(0, server_1.makeHTMLParts)(),
                ...htmlPartsFromConfig,
            };
            await hooks.process("ruruHTMLParts", htmlParts, { request });
        }
        const config = {
            endpoint: dynamicOptions.graphqlPath,
            // TODO: websocket endpoint
            debugTools: dynamicOptions.explain === true
                ? ["explain", "plan"]
                : dynamicOptions.explain === false
                    ? []
                    : dynamicOptions.explain,
        };
        return {
            statusCode: 200,
            request,
            dynamicOptions,
            type: "html",
            payload: Buffer.from((0, server_1.ruruHTML)(config, htmlParts), "utf8"),
        };
    };
}
exports.makeGraphiQLHandler = makeGraphiQLHandler;
