// graphile-build-pg/dist/pgServices.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withSuperuserPgClientFromPgService = exports.withPgClientFromPgService = exports.getWithPgClientFromPgService = void 0;
const grafast_1 = require("grafast");
const isTest = process.env.NODE_ENV === "test";
const withPgClientDetailsByConfigCache = new Map();
function reallyLoadAdaptor(adaptorString) {
    try {
        const adaptor = require(adaptorString);
        return adaptor?.createWithPgClient ? adaptor : adaptor?.default;
    }
    catch (e) {
        if (e.code === "ERR_REQUIRE_ESM") {
            const adaptorPromise = import(adaptorString);
            return adaptorPromise.then((adaptor) => adaptor?.createWithPgClient ? adaptor : adaptor?.default);
        }
        else {
            throw e;
        }
    }
}
const loadAdaptorCache = new Map();
function loadAdaptor(adaptorString) {
    const cached = loadAdaptorCache.get(adaptorString);
    if (cached) {
        return cached;
    }
    else {
        const result = reallyLoadAdaptor(adaptorString);
        loadAdaptorCache.set(adaptorString, result);
        if ((0, grafast_1.isPromiseLike)(result)) {
            result.then((resolved) => {
                loadAdaptorCache.set(adaptorString, resolved);
            }, () => { });
        }
        return result;
    }
}
/**
 * Get or build the 'withPgClient' callback function for a given database
 * config, caching it to make future lookups faster.
 */
function getWithPgClientFromPgService(config) {
    const existing = withPgClientDetailsByConfigCache.get(config);
    if (existing) {
        if ((0, grafast_1.isPromiseLike)(existing)) {
            return existing.then((v) => {
                v.retainers++;
                return v.withPgClient;
            });
        }
        else {
            existing.retainers++;
            return existing.withPgClient;
        }
    }
    else {
        const promise = (async () => {
            const adaptor = await loadAdaptor(config.adaptor);
            const factory = adaptor?.createWithPgClient;
            if (typeof factory !== "function") {
                throw new Error(`'${config.adaptor}' does not look like a withPgClient adaptor - please ensure it exports a method called 'createWithPgClient'`);
            }
            const originalWithPgClient = await factory(config.adaptorSettings);
            const withPgClient = ((...args) => originalWithPgClient.apply(null, args));
            const cachedValue = {
                withPgClient,
                retainers: 1,
            };
            let released = false;
            withPgClient.release = () => {
                cachedValue.retainers--;
                // To allow for other promises to resolve and add/remove from the retaininers, check after a tick
                setTimeout(() => {
                    if (cachedValue.retainers === 0 && !released) {
                        released = true;
                        withPgClientDetailsByConfigCache.delete(config);
                        return originalWithPgClient.release?.();
                    }
                    // TODO: this used to be zero, but that seems really inefficient...
                    // Figure out why I did that?
                    // }, 0);
                }, isTest ? 500 : 5000);
            };
            withPgClientDetailsByConfigCache.set(config, cachedValue);
            return cachedValue;
        })();
        if (!withPgClientDetailsByConfigCache.has(config)) {
            withPgClientDetailsByConfigCache.set(config, promise);
        }
        promise.catch(() => {
            withPgClientDetailsByConfigCache.delete(config);
        });
        return promise.then((v) => v.withPgClient);
    }
}
exports.getWithPgClientFromPgService = getWithPgClientFromPgService;
async function withPgClientFromPgService(config, pgSettings, callback) {
    const withPgClientFromPgService = getWithPgClientFromPgService(config);
    const withPgClient = (0, grafast_1.isPromiseLike)(withPgClientFromPgService)
        ? await withPgClientFromPgService
        : withPgClientFromPgService;
    try {
        return await withPgClient(pgSettings, callback);
    }
    finally {
        withPgClient.release();
    }
}
exports.withPgClientFromPgService = withPgClientFromPgService;
// We don't cache superuser withPgClients
async function withSuperuserPgClientFromPgService(config, pgSettings, callback) {
    const adaptor = await loadAdaptor(config.adaptor);
    const withPgClient = await adaptor.createWithPgClient(config.adaptorSettings, "SUPERUSER");
    try {
        return await withPgClient(pgSettings, callback);
    }
    finally {
        withPgClient.release?.();
    }
}
exports.withSuperuserPgClientFromPgService = withSuperuserPgClientFromPgService;
